### ISSUE TITLE
[Meta] Backwards compatibility and deprecation warnings

### ISSUE THREAD

**Issue by HaleTom** [tmux/tmux#1732]
### Context

This is a **meta**-issue about this project's philosophy of introducing non-backward compatible changes in minor (eg the `Y` in `vX.Y` ) releases.

Individual instances of the pain associated with this philosophy in relation to specific releases have been raised:
* [2.1: Missing backwards compatibility #155](https://github.com/tmux/tmux/issues/155)
* [Tmux 2.8.X to 2.9.X migration #1689](https://github.com/tmux/tmux/issues/1689)

But this issue is not about any particular release's non-backward-compatibility. It is about the strategy of:

1. Generally not allowing for backward-compatibility
2. Having breaking changes released as minor versions

A reasonable expectation in regards to breaking changes is outlined by [semver](https://semver.org/):

> Given a version number MAJOR.MINOR.PATCH, increment the:
>
> 1. MAJOR version when you make incompatible API changes,
> 2. MINOR version when you add functionality in a backwards-compatible manner, and
> 3. PATCH version when you make backwards-compatible bug fixes.

I recognise that this project has chosen it's own versioning, but I can't the definition of what is implied by different types of version bumps.

Please note that I'm **not** suggesting that breaking minor changes should have been a major version bump instead.

I **am** proposing that backward compatibility is considered **usual** and **best practice**.

### Issue
As an example of what occurs on every breaking change, I timed how long it took me to read the 2.8 -> 2.9 issue (Google took me there first), read the doco and examples, and update based on the recent `-style` changes: 23 minutes.

I consider myself a medium-to-advanced level of proficiency. If every starred user takes 20 minutes to do the update, then that's:

Total: 4601 hours or **192 days** of human time.  And that's *only* the starred users. It's probably way more than 10x that, or a **conservative 5.25 human years**.

I use the rolling-release Arch Linux which provides the latest `tmux`, which in turn **requires** me to have a `.tmux.conf` which aligns to the latest version of tmux.

What am I supposed to do when I want to use a fixed-release distribution such as Ubuntu with an older version of `tmux`?

### Work-around
To support older LTS distributions as well as my more more current workstation I am forced to either:

1. Maintain multiple `.tmux.conf` files
1. [Use an ugly hack to allow my config to work across multiple versions](https://stackoverflow.com/a/40902312/5353461):

```
# Version-specific commands [grumble, grumble]
# See: https://github.com/tmux/tmux/blob/master/CHANGES
run-shell "tmux setenv -g TMUX_VERSION $(tmux -V | cut -c 6-)"

if-shell -b '[ "$(echo "$TMUX_VERSION < 2.1" | bc)" = 1 ]' " \
    set -g mouse-select-pane on; set -g mode-mouse on; \
    set -g mouse-resize-pane on; set -g mouse-select-window on \
    set -g message-fg red \
    set -g message-bg black \
    set -g message-attr bright \
    set -g window-status-bg default \
    set -g window-status-fg default \
    set -g window-status-current-attr bold \
    set -g window-status-current-bg cyan \
    set -g window-status-current-fg default \
    set -g window-status-bell-fg red \
    set -g window-status-bell-bg black \
    set -g window-status-activity-fg white \
    set -g window-status-activity-bg black"

# In version 2.1 "mouse" replaced the previous 4 mouse options
if-shell -b '[ "$(echo "$TMUX_VERSION >= 2.1" | bc)" = 1 ]' \
  "set -g mouse on"

# UTF8 is autodetected in 2.2 onwards, but errors if explicitly set
if-shell -b '[ "$(echo "$TMUX_VERSION < 2.2" | bc)" = 1 ]' \
  "set -g utf8 on; set -g status-utf8 on; set -g mouse-utf8 on"

# bind-key syntax changed in 2.4
if-shell -b '[ "$(echo "$TMUX_VERSION < 2.4" | bc)" = 1 ]' \
  "# Selection / copy / paste \
   bind-key -t vi-copy v   begin-selection; \
   bind-key -t vi-copy V   send -X select-line \
   bind-key -t vi-copy C-v rectangle-toggle \
   bind-key -t vi-copy y   copy-pipe 'xclip -selection clipboard -in'"

# Newer versions
if-shell -b '[ "$(echo "$TMUX_VERSION < 2.9" | bc)" = 1 ]' \
  "# Selection / copy / paste \
   bind-key -T copy-mode-vi v   send -X begin-selection  # default is <space> \
   bind-key -T copy-mode-vi V   send -X select-line \
   bind-key -T copy-mode-vi C-v send -X rectangle-toggle  # default is C-v, or R in copy-mode (non-vi) \
   bind-key -T copy-mode-vi y   send -X copy-pipe-and-cancel 'xclip -selection clipboard -in'"

if-shell -b '[ "$(echo "$TMUX_VERSION >= 2.9" | bc)" = 1 ]' \
   "set -g message-style fg=red,bg=black \
    set -g message-style bright \
    set -g window-status-style          fg=default,bg=default \
    set -g window-status-current-style  fg=default,bg=cyan,bold \
    set -g window-status-bell-style     fg=red,bg=black \
    set -g window-status-activity-style fg=white,bg=black"
```

The issue with using the hack above:

1. It's ugly
1. It loses syntax highlighting in my editor (due to the long multi-line strings)
1. It means that  part of my config is organised by version, not by functionality
1. It means that to properly fudge backward compatibilty, I need to look at the `CHANGES` file for the introduction version of each directive, to make sure I add the directive to the correct version section.

Still, my horrible kludge has [33 upvotes on StackOverflow](https://stackoverflow.com/a/40902312/5353461).

(Note my 23 minutes to work around the v2.9 breaking changes *did not* include the extra time taken to include both 2.9 and earlier config lines into my workaround kludge structure)

### Proposal

I don't usually use Ubuntu, but when I do, I'm pretty much guaranteed to hit an old version of `tmux`.

Ubuntu LTS support is 3 years.  Other common distributions supported by `tmux` may have longer support lives.

I propose that `tmux` chooses a support window based upon either Ubuntu or another common distribution which may have a longer support life.

### TL;DR: Summary

I propose that this project takes on a **minimum** 3 year backward-compatibility support window for releases.

**Comment by nicm** [comment]
-style options were introduced in tmux 1.9 in 2015 so you have had plenty of time to get used to them and they should be available in any version that should still be around.

Removing and adding options should not be a big problem since set-option -q was added in tmux 1.7 in 2012. 

I admit that the CHANGES file and the documentation was too bare for -style. But tmux 2.9 had multiple release candidates over one month - that was the time to mention it, that was the time that volunteering to write a migration script or document would have been helpful. But nobody did.

I don't have particular interest in managing a formal process for deprecating stuff. And without meaning to be rude - it seems like it would be more helpful if rather than telling me what I should do, you suggest what YOU could do?

Any features to help with this - such as set -q, %if, #{version} (both added in 2.4 - 2 years ago now) - have been largely driven by me. People have not been sending patches or even feature requests saying "backwards compatibility would be easier if we did this", even after previous changes. If nobody mentions it, you will just get what I want - and I build a new tmux on old machines, so backwards compatibility is not top of my list.

**Comment by nicm** [comment]
Here are some suggestions for things people could do if they felt so inclined. tmux is not a large project so none of this would take much time, in no particular order:

- Go through the git commits either during development or before release and highlight any breaking changes so we can decide on their impact.

- Review the CHANGES file before release for completeness and a accuracy. I know how the changes work so it's easy to explain them too barely.

- Work on migration, upgrade, new feature documents for releases if necessary.

- Improve the FAQ. Write (and maintain!) some official howto guides.

- Help to tag and release rolling RCs through the year, or manage a longer RC process, or help to announce RCs more widely, or just test them.

- Help with packaging RCs for different platforms or produce static builds so people can test more easily.

- Maintain one or more stable tmux branches and curate changes in from master.

- Write regression tests. Test master or RCs and report bugs.

**Comment by HaleTom** [comment]
Firstly, @nicm, thanks for all the work that you've been putting in to this project. I guess that we'll never know how many people use it, but you're probably making the lives of hundreds of thousands of people easier.

I'll reply about backwards compatibility later, but for now, would it assist if I make your list into a "[meta] Suggested contributions" issue so that people looking to contribute have a pointer as to where to start?

**Comment by nicm** [comment]
I think it might be better as a web page than an issue, although many of these are long term and just need one person to take them on.

**Comment by nicm** [comment]
Wiki page even. The best bits from the TODO file could go on there too.

**Comment by HaleTom** [comment]
From a bit of a google, it seems that it's not possible to make PRs on wiki pages.

A copy & paste to a "Suggested contributions" wiki page would ensure that these points don't get lost - unresolved yet hidden away in this issue.

**Comment by HaleTom** [comment]
I'm sad that you took my words as `telling me what I should do` rather than as a proposal for what this project could do to make the lives of its users easier.

Thanks for the reminders about `set -q`  and `%if #{version}` added in version 2.4.  Using these features will definitely improve the readability of my `.tmux.conf`.

However, for me and many others, it will be premature to use these v2.4 features until after [Ubuntu 16.04's maintenance ends in April 2021](https://www.ubuntu.com/about/release-cycle) as [Ubuntu 16.04 has tmux v2.1-3build1](https://packages.ubuntu.com/search?keywords=tmux).  I look forward to updating my config some time after that, and hope to run into Ubuntu 16.04 as few times as possible in the interim.

You suggested many excellent things that could be done around release management, but I do feel you missed my main point. 

My main point was about the tens of thousands of human lives and over 5 years of human life that is affected every time that the preexisting format of the config file is no longer supported. I ask this project's maintainers to look into what 5 years of your life means to you, and whether creating a non-backward-compatible change is worth affecting that many human lives.

There may be rare times when it is *necessary* to not allow a old configuration file to continue working, but in my 30 years of experience with config files, this is indeed *rare*.

My further suggestions are:

* To maximise backward compatibility (minimise breaking change) in the code so as to reduce the release management impact
* Where breaking change (non-backward-compatible change) is considered necessary: to give warning (ideally for a minimum defined period of time) of any newly deprecated and soon-to-be unsupported syntax.

Please consider the effort that I've put into these suggestions as a contribution to this project in the larger sense - as a contribution to (and feedback from) the ecosystem that the project exists in (users using it), and with the intent to reduce impact and frustration in people's lives, when we are already trying to do the best we can with the challenges we already face.

@nicm I see that you are the name mentioned at the bottom of the `man`-page. What you've built has already reduced the frustration in my life, and `tmux` is an invaluable tool. I hope this project contnues to foster teamwork to lighten your load, as reading between the lines, a lot of it seems to fall upon you. 
Again, thank you. Even if you take up none of what I suggest, you have already made a positive impact to my quality of life.

**Comment by nicm** [comment]
I do appreciate your comments but the fact is that although suggestions are all good and well,  someone has to do the work.

I don't have the time or interest in tracking what tmux versions are available on various LTS platforms that I don't use.


I don't want to keep code in tmux for backwards compatibility indefinitely. It is a maintenance burden and not how I want to spend my time. It has to go eventually and how long to wait is largely judged by how much impact I perceive - unless people tell me otherwise. This change waited for almost four years. That seems long enough.

I considered adding a warning but  there would have been no point - the same config file changes would be needed, the warning would have looked the same as the errors do. And remember this wasn't a breaking change, tmux still started and worked, just with perhaps the wrong colours in the status line. This is only perhaps the fourth or fifth change that affected the config file in ten years.

Saying that, I'm not unsympathetic and if someone had pointed out to me that this change affected Ubuntu LTS users or whatever and suggested deferring it for a later release, I would have listened.

tmux is a small program. It has hardly any dependencies and takes no time to build. It's target audience is heavy terminal users - developers, sysadmins, Unix nerds. I don't think it is too much to ask that people test and speak up before the releases, or manage the occasional incompatibility or bug if they don't.

**Comment by krader1961** [comment]
> Saying that, I'm not unsympathetic and if someone had pointed out to me that this change affected Ubuntu LTS users or whatever and suggested deferring it for a later release, I would have listened.

@nicm, FWIW, I sometimes find your response to an issue overly terse or dismissive. Having said that I think you have shown great restraint this time. I know I would have taken a break to perform deep breathing exercises to calm down, and forcing myself to go through a sleep cycle, before responding as you did.

I spent a lot of my life working for a computer vendor and thus appreciate the need to maximize backward compatibility. But that is an unrealistic goal for a project with no commercial support and very few people who contribute changes on a regular basis. In this instance I believe you made a reasonable trade-off; i.e., compromise. It is difficult to accommodate LTS (long term support) platforms. The one thing I might have done differently is to make it crystal clear a release had backward incompatible changes by incrementing the major release number.

**Comment by nicm** [comment]
Yes, I know I can be too terse sometimes - it is one or all of trying to be concise, being busy, replying on a phone, or sometimes being annoyed when people don't even really try to figure out things themselves. But I try to be patient.

I do understand that it is annoying when config syntax changes between releases and if I had known ahead of time I would have made this more obvious. But do I think this needs to be kept in perspective:

- tmux is a small project, it is easy to build - we are not talking about KDE here;
- it has one text config file, nobody is being asked to run gnome-settings --fiddle and find some obscure corner of the UI, or dig around in a load of *.el files;
- -style was introduced four years ago, that is almost half of tmux's life;
- -fg/-bg/-attr was removed from the manual and not shown by show-options at the same time;
- for users who started using tmux before 1.9, the CHANGES file for 1.9 release says the old options would be removed in time;
- 2.9 also has a new STYLES section of the manual which more obviously documents the syntax;
- mechanisms to support old versions in .tmux.conf ARE available - it's unfortunate they weren't around earlier but it's too late to do anything about that now;
- nobody is forced to upgrade - if you still need to use tmux 1.8 in Ubuntu LTS, stick with tmux 2.8 until the next Ubuntu release.

So I don't think it was silly to remove them at this point. Things have to progress or you end up with a unmaintainable codebase full of kludges for obscure features and ancient versions that nobody uses.

I should have made more noise about it but it's easy to say that in hindsight - nobody asked about them in the couple of months before release (or in the last four years), so it wasn't unreasonable not to worry about it.

I use tmux in one way, I don't have time to track the way everyone else uses it and what options and versions they are using and on what platforms. I try to keep things working but by and large I have to rely on users to tell me if there are problems. People often do, and I am grateful for that, but when they don't there isn't always a lot I can do about it.

I don't like feature versioning because it is so subjective, one person's major feature is another's minor, and they easily end up being silly (either 0.999.2 or 999.2.0). And I don't really work like that, I write major features when I have the time, I don't target them to a release. A steadily ticking version number is more predictable.

**Comment by nicm** [comment]
In any case, I don't think there is much mileage in discussing this much further.

I don't intend to avoid configuration file or behaviour changes entirely or to change the version numbering, but I will try to document changes more thoroughly.

I encourage all users to test tmux release candidates or Git master if possible and let me know of problems.

If anyone would like to take on any of the suggestions I have made above, or any others, of course please get in touch. I'll move them to a separate page at some point but I'd like to move the TODO file as well.

**Comment by HaleTom** [comment]
@nicm You said:

> And remember this wasn't a breaking change, tmux still started and worked, just with perhaps the wrong colours in the status line. 

In a few organisations I have raised Change Requests and sat on Change Review Boards, and from my experience your definition of "non-breaking" is quite peculiar:

Functionality which previously worked no longer worked.

While it's not my use case, I have seen one company provide a template where the Dev / Test / Prod terminal background colours were set differently to reduce the risk of stuff-ups.

With so many users, it's impossible to tell the real-world impact of (in my terms) a breaking change.

> I don't want to keep code in tmux for backwards compatibility indefinitely. [...] This change waited for almost four years. That seems long enough.

I appreciate that backward compatibility carries technical debt. 

It's great that you kept this backward compatible for 4 years. I didn't realise that. Thank you!

> I considered adding a warning but there would have been no point - the same config file changes would be needed, the warning would have looked the same as the errors do.

If there were depreciation warnings, I could have ignored the startup warnings and *retained my existing working environment* for a few weeks/months before finding a spare "round tuit" to silence the warnings.

I also noticed the different tone in your reply, and believe that it will welcome more contributions of the more valuable PR-type in the future.

Thanks for your time in replying.

**Comment by nicm** [comment]
Right, but my point is that if you want to use tmux immediately, having the wrong colours in the status line is a minor problem.

> If there were depreciation warnings, I could have ignored the startup warnings and retained my existing working environment for a few weeks/months before finding a spare "round tuit" to silence the warnings.

Perhaps, but you could equally have waited until you had time to read through the CHANGES file and fix up your config before upgrading.

It should certainly have been more visibly documented in CHANGES, but it was there and in the release announcement email.

**Comment by ThomasAdam** [comment]
It's always nice to hear from people who aren't happy with certain situations -- it's reaffirming that the software being produced matters, all the while, we get to listen to rhetoric.

I'll be going back to performing release-candidate/release duties for tmux and will start to make a lot more noise about any "breaking" changes.

However, please note that:

* There won't be any _code_ changes to implement deprecation warnings -- this is as much a burden as it is keeping backwards compatibility.

* Despite the current observations, this is the first time there has been this many config file changes in one release.  The others went by largely unnoticed.

Here's to a gentler tmux-3.0 and beyond.

**Comment by HaleTom** [comment]
@nicm said:

> Perhaps, but you could equally have waited until you had time to read through the CHANGES file and fix up your config before upgrading.

That's a *very* unlikely situation for the majority of users.

`pacman -Syu` doesn't show me the `CHANGES` files of all the packages it will upgrade.

Likewise, users of non-rolling-release distros will only learn about breaking changes... when they break.

@ThomasAdam I accept that there will be no deprecation warnings in code.

Can you please notify users of a version numbering strategy (semver is pre-baked) which highlights non-backward-compatible changes?

`pacman` does show version numbers when upgrading.

**Comment by ThomasAdam** [comment]
@HaleTom there is no version numbering strategy, other than to keep incrementing forwards for each release, which is approximately every six months, per the OpenBSD release cycle.

Speaking of which, I don't want to break away from that to something as OTT (for tmux) as semver. In the history of tmux portable releases, there's been maybe three releases on top of a typical release, which was an 'emergency' release fix. That's a good testimony to why semver isn't needed here. Tmux is an OpenBSD project, after all.

As we've said all along, keep reading CHANGES.

**Comment by nicm** [comment]
> That's a very unlikely situation for the majority of users.
>
> pacman -Syu doesn't show me the CHANGES files of all the packages it will upgrade.
>
> Likewise, users of non-rolling-release distros will only learn about breaking changes... when they break.

I think this is all getting a bit all over the place really. I think I am mostly just repeating myself now so I will stop after this reply.

I don't have much sympathy for anyone who is happy to fire off an upgrade when they won't have 20 minutes spare to deal with any fallout. Even if the tmux config hadn't changed, there could still have been annoying bugs, or packaging issues, or a dozen other kinds of problem. That is the price of rolling releases.

If the package system doesn't tell you about package changes or the maintainer chooses not to, that's a choice they have made.

Warnings are not free, it takes time to write the code and there is a maintenance cost. As I said, I considered them for this, but I didn't think they were worth it. The changes are obvious as soon as you start tmux and they did not fundamentally stop tmux working. Agree or disagree, it is too late now.

We aren't changing the versioning scheme, but we will make stuff like this more visible in the CHANGES file, I do agree that it was not highlighted well enough. But this is why users helping to test and report issues in master and during the release process is important, not just bugs but problems with the way changes and new features are documented.

I do sympathize with the fact that it can be hard to use the same .tmux.conf with multiple versions of tmux simultaneously. But I have worked on this - there are now a number of ways to help deal with it and there is not much we can do about the fact they are not available in some of the versions out there except wait.

The fact is that I have limited time and I am not going to spend too much of it worrying about releases and process at the expense of the stuff that I enjoy doing - writing code and helping people with issues and questions. Thomas is going to take over the releases again so perhaps he will have more time.

**Comment by HaleTom** [comment]
Cheers @ncim and @ThomasAdam.
